cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -0,0 +1,198 @@
+        String dag2Name = subdagEntry2.getKey();
+        Dag dag2 = subdagEntry2.getValue();
+        if (dag1Name.equals(dag2Name)) {
+          continue;
+        }
+
+        // if dag1 has any sinks that are a source in dag2, add a connection between the dags
+        if (Sets.intersection(dag1.getSinks(), dag2.getSources()).size() > 0) {
+          phaseConnections.add(new Connection(dag1Name, dag2Name));
+        }
+      }
+    }
+
+    // convert to objects the programs expect.
+    Map<String, PipelinePhase> phases = new HashMap<>();
+    for (Map.Entry<String, Dag> dagEntry : subdags.entrySet()) {
+      phases.put(dagEntry.getKey(), dagToPipeline(dagEntry.getValue(), connectorNodes, specs));
+    }
+    return new PipelinePlan(phases, phaseConnections);
+  }
+
+  /**
+   * Converts a Dag into a PipelinePhase, using what we know about the plugin type of each node in the dag.
+   * The PipelinePhase is what programs will take as input, and keeps track of sources, transforms, sinks, etc.
+   *
+   * @param dag the dag to convert
+   * @param connectors connector nodes across all dags
+   * @param specs specifications for every stage
+   * @return the converted dag
+   */
+  private PipelinePhase dagToPipeline(Dag dag, Set<String> connectors, Map<String, StageSpec> specs) {
+    Set<StageInfo> sinks = new HashSet<>();
+    Set<StageInfo> transforms = new HashSet<>();
+    Map<String, Set<String>> connections = new HashMap<>();
+    StageInfo source = null;
+    StageInfo aggregator = null;
+
+    for (String stageName : dag.linearize()) {
+      Set<String> outputs = dag.getNodeOutputs(stageName);
+      if (!outputs.isEmpty()) {
+        connections.put(stageName, dag.getNodeOutputs(stageName));
+      }
+
+      // if its a connector, figure out if its a source or a sink
+      if (connectors.contains(stageName)) {
+        StageInfo stageInfo = new StageInfo(stageName, null, true);
+        if (dag.getSources().contains(stageName)) {
+          source = stageInfo;
+        } else {
+          sinks.add(stageInfo);
+        }
+        continue;
+      }
+
+      StageSpec spec = specs.get(stageName);
+      String pluginType = spec.getPlugin().getType();
+
+      StageInfo stageInfo = new StageInfo(stageName, spec.getErrorDatasetName(), connectors.contains(stageName));
+      if (sinkType.equals(pluginType)) {
+        sinks.add(stageInfo);
+      } else if (Transform.PLUGIN_TYPE.equals(pluginType)) {
+        transforms.add(stageInfo);
+      } else if (sourceType.equals(pluginType)) {
+        if (source != null) {
+          // should never happen
+          throw new IllegalStateException(String.format(
+            "There was an error during pipeline planning. " +
+              "A pipeline phase should not have multiple sources, but %s and %s are in the same phase.",
+            source.getName(), stageName));
+        }
+        source = stageInfo;
+      } else if ("aggregator".equals(pluginType)) {
```

Comment:
this is a placeholder until the actual aggregator class is added.  Will replace it with the constant when that happens.